### PR TITLE
[core] Bring cpu device module closer to cuda's.

### PR DESCRIPTION
**Pull request by kumpera**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103172

By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc @ptrblck @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: torch/cpu/__init__.py [hunk 1]
```diff
@@ -1 +1,85 @@
...
+    """
+    return _current_stream
+
+class StreamContext:
+    r"""Context-manager that selects a given stream.
+
+    N.B. This class only exists to facilitate device-agnostic code
+
+    """
+    cur_stream : Optional['torch.cuda.Stream']
```
Comment (wz337) [inline torch/cpu/__init__.py:49]: nit: import torch?

File: torch/cpu/__init__.py [hunk 2]
```diff
@@ -1,6 +1,90 @@
+from typing import Any, Optional
 from . import amp
 import torch
 
 def _is_cpu_support_vnni() -> bool:
     r"""Returns a bool indicating if CPU supports VNNI."""
     return torch._C._cpu._is_cpu_support_vnni()
+
+
+def is_available() -> bool:
```
Comment (wanchaol) [inline torch/cpu/__init__.py:10]: shall we also add `set_device`?

File: torch/cpu/__init__.py (hunk 1 above)
Comment (kumpera) [reply torch/cpu/__init__.py:54]: can't, it causes an import loop :(

File: docs/source/cpu.rst [hunk 3]
```diff
@@ -0,0 +1,23 @@
+torch.cuda
```
Comment (wz337) [inline docs/source/cpu.rst:1]: nit: This should be torch.cpu?

**Comment by wz337** [comment]
Noob question: is there any way to add a test for this?

**Review by wz337 — APPROVED** [review]
LGTM

File: docs/source/cpu.rst (hunk 3 above)
Comment (kumpera) [reply docs/source/cpu.rst:1]: yup

**Review by wanchaol — APPROVED** [review]
nit: `set_device` is still not there so it might not be a fully device agnositic yet? I think `torch.cpu.set_device` should be accepting a integer similar to cuda but does nothing inside.

**Comment by albanD** [comment]
FYI @heidongxianhua that should allow us to harmonize all the backend code in the different integrations we're adding!

File: torch/cpu/__init__.py [hunk 4]
```diff
@@ -1,6 +1,109 @@
+r"""
+This package implements abstractions found in ``torch.cuda``
```
Comment (albanD) [inline torch/cpu/__init__.py:2]: Note that this is missing a few more things from the list at https://github.com/pytorch/pytorch/blob/999abd56a7be3d73f0d2e9ab06356c7281a3307d/torch/utils/backend_registration.py#L31

But that can be done in a follow up for sure.

Comment (kumpera) [reply torch/cpu/__init__.py:2]: yes, I plan to follow this up with Protocol'ization of this design so we can have tests for this.

**Pull request closed without merging**